// immutable-json-patch/lib/cjs/immutabilityHelpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.shallowClone = shallowClone;
exports.applyProp = applyProp;
exports.getIn = getIn;
exports.setIn = setIn;
exports.updateIn = updateIn;
exports.deleteIn = deleteIn;
exports.insertAt = insertAt;
exports.transform = transform;
exports.existsIn = existsIn;

var _utils = require("./utils.js");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

/**
 * Shallow clone of an Object, Array, or value
 * Symbols are cloned too.
 * @param {*} value
 * @return {*}
 */
function shallowClone(value) {
  if (Array.isArray(value)) {
    // copy array items
    var copy = value.slice(); // copy all symbols

    Object.getOwnPropertySymbols(value).forEach(function (symbol) {
      copy[symbol] = value[symbol];
    });
    return copy;
  } else if (_typeof(value) === 'object') {
    // copy object properties
    var _copy = _objectSpread({}, value); // copy all symbols


    Object.getOwnPropertySymbols(value).forEach(function (symbol) {
      _copy[symbol] = value[symbol];
    });
    return _copy;
  } else {
    return value;
  }
}
/**
 * Update a value in an object in an immutable way.
 * If the value is unchanged, the original object will be returned
 * @param {Object | Array} object
 * @param {string | index} key
 * @param {*} value
 * @returns {Object | Array}
 */


function applyProp(object, key, value) {
  if (object[key] === value) {
    // return original object unchanged when the new value is identical to the old one
    return object;
  } else {
    var updatedObject = shallowClone(object);
    updatedObject[key] = value;
    return updatedObject;
  }
}
/**
 * helper function to get a nested property in an object or array
 *
 * @param {Object | Array} object
 * @param {JSONPath} path
 * @return {* | undefined} Returns the field when found, or undefined when the
 *                         path doesn't exist
 */


function getIn(object, path) {
  var value = object;
  var i = 0;

  while (i < path.length) {
    if ((0, _utils.isObjectOrArray)(value)) {
      value = value[path[i]];
    } else {
      value = undefined;
    }

    i++;
  }

  return value;
}
/**
 * helper function to replace a nested property in an object with a new value
 * without mutating the object itself.
 *
 * @param {Object | Array} object
 * @param {JSONPath} path
 * @param {*} value
 * @param {boolean} [createPath=false]
 *                    If true, `path` will be created when (partly) missing in
 *                    the object. For correctly creating nested Arrays or
 *                    Objects, the function relies on `path` containing a number
 *                    in case of array indexes.
 *                    If false (default), an error will be thrown when the
 *                    path doesn't exist.
 * @return {Object | Array} Returns a new, updated object or array
 */


function setIn(object, path, value) {
  var createPath = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : false;

  if (path.length === 0) {
    return value;
  }

  var key = path[0];
  var updatedValue = setIn(object ? object[key] : undefined, path.slice(1), value, createPath);

  if (!(0, _utils.isObjectOrArray)(object)) {
    if (createPath) {
      var newObject = typeof key === 'number' ? [] : {};
      newObject[key] = updatedValue;
      return newObject;
    } else {
      throw new Error('Path does not exist');
    }
  }

  return applyProp(object, key, updatedValue);
}
/**
 * helper function to replace a nested property in an object with a new value
 * without mutating the object itself.
 *
 * @param {Object | Array} object
 * @param {JSONPath} path
 * @param {function} callback
 * @return {Object | Array} Returns a new, updated object or array
 */


function updateIn(object, path, callback) {
  if (path.length === 0) {
    return callback(object);
  }

  if (!(0, _utils.isObjectOrArray)(object)) {
    throw new Error('Path doesn\'t exist');
  }

  var key = path[0];
  var updatedValue = updateIn(object[key], path.slice(1), callback);
  return applyProp(object, key, updatedValue);
}
/**
 * helper function to delete a nested property in an object
 * without mutating the object itself.
 *
 * @param {Object | Array} object
 * @param {JSONPath} path
 * @return {Object | Array} Returns a new, updated object or array
 */


function deleteIn(object, path) {
  if (path.length === 0) {
    return object;
  }

  if (!(0, _utils.isObjectOrArray)(object)) {
    throw new Error('Path does not exist');
  }

  if (path.length === 1) {
    var _key = path[0];

    if (!(_key in object)) {
      // key doesn't exist. return object unchanged
      return object;
    } else {
      var updatedObject = shallowClone(object);

      if (Array.isArray(updatedObject)) {
        updatedObject.splice(_key, 1);
      } else {
        delete updatedObject[_key];
      }

      return updatedObject;
    }
  }

  var key = path[0];
  var updatedValue = deleteIn(object[key], path.slice(1));
  return applyProp(object, key, updatedValue);
}
/**
 * Insert a new item in an array at a specific index.
 * Example usage:
 *
 *     insertAt({arr: [1,2,3]}, ['arr', '2'], 'inserted')  // [1,2,'inserted',3]
 *
 * @param {Object | Array} object
 * @param {JSONPath} path
 * @param {*} value
 * @return {Array}
 */


function insertAt(object, path, value) {
  var parentPath = path.slice(0, path.length - 1);
  var index = path[path.length - 1];
  return updateIn(object, parentPath, function (items) {
    if (!Array.isArray(items)) {
      throw new TypeError('Array expected at path ' + JSON.stringify(parentPath));
    }

    var updatedItems = shallowClone(items);
    updatedItems.splice(index, 0, value);
    return updatedItems;
  });
}
/**
 * Transform a JSON object, traverse over the whole object,
 * and allow replacing Objects/Arrays/values.
 * @param {JSONData} json
 * @param {function (json: JSONData, path: JSONPath) : JSONData} callback
 * @param {JSONPath} [path]
 * @return {JSONData}
 */


function transform(json, callback) {
  var path = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];
  var updated1 = callback(json, path);

  if (Array.isArray(json)) {
    // array
    var updated2;

    for (var i = 0; i < updated1.length; i++) {
      var before = updated1[i]; // we stringify the index here, so the path only contains strings and can be safely
      // stringified/parsed to JSONPointer without loosing information.
      // We do not want to rely on path keys being numeric/string.

      var after = transform(before, callback, path.concat(i + ''));

      if (after !== before) {
        if (!updated2) {
          updated2 = shallowClone(updated1);
        }

        updated2[i] = after;
      }
    }

    return updated2 || updated1;
  } else if (json && _typeof(json) === 'object') {
    // object
    var _updated;

    for (var key in updated1) {
      if (Object.hasOwnProperty.call(updated1, key)) {
        var _before = updated1[key];

        var _after = transform(_before, callback, path.concat(key));

        if (_after !== _before) {
          if (!_updated) {
            _updated = shallowClone(updated1);
          }

          _updated[key] = _after;
        }
      }
    }

    return _updated || updated1;
  } else {
    // number, string, boolean, null
    return updated1;
  }
}
/**
 * Test whether a path exists in a JSON object
 * @param {JSONData} json
 * @param {JSONPath} path
 * @return {boolean} Returns true if the path exists, else returns false
 * @private
 */


function existsIn(json, path) {
  if (json === undefined) {
    return false;
  }

  if (path.length === 0) {
    return true;
  }

  return existsIn(json[path[0]], path.slice(1));
}
