// immutable-json-patch/lib/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "immutableJSONPatch", {
  enumerable: true,
  get: function get() {
    return _immutableJSONPatch.immutableJSONPatch;
  }
});
Object.defineProperty(exports, "revertJSONPatch", {
  enumerable: true,
  get: function get() {
    return _revertJSONPatch.revertJSONPatch;
  }
});
Object.defineProperty(exports, "compileJSONPointer", {
  enumerable: true,
  get: function get() {
    return _jsonPointer.compileJSONPointer;
  }
});
Object.defineProperty(exports, "parseJSONPointer", {
  enumerable: true,
  get: function get() {
    return _jsonPointer.parseJSONPointer;
  }
});
Object.defineProperty(exports, "getIn", {
  enumerable: true,
  get: function get() {
    return _immutabilityHelpers.getIn;
  }
});
Object.defineProperty(exports, "setIn", {
  enumerable: true,
  get: function get() {
    return _immutabilityHelpers.setIn;
  }
});
Object.defineProperty(exports, "existsIn", {
  enumerable: true,
  get: function get() {
    return _immutabilityHelpers.existsIn;
  }
});
Object.defineProperty(exports, "insertAt", {
  enumerable: true,
  get: function get() {
    return _immutabilityHelpers.insertAt;
  }
});
Object.defineProperty(exports, "deleteIn", {
  enumerable: true,
  get: function get() {
    return _immutabilityHelpers.deleteIn;
  }
});
Object.defineProperty(exports, "updateIn", {
  enumerable: true,
  get: function get() {
    return _immutabilityHelpers.updateIn;
  }
});

var _immutableJSONPatch = require("./immutableJSONPatch.js");

var _revertJSONPatch = require("./revertJSONPatch.js");

var _jsonPointer = require("./jsonPointer.js");

var _immutabilityHelpers = require("./immutabilityHelpers.js");
